moveit_core/robot_state/test/robot_state_benchmark.cpp

```diff
@@ -66,312 +61,279 @@ Eigen::Isometry3d createTestIsometry()
+  Eigen::Isometry3d result;
+  for (auto _ : st)
+  {
+    benchmark::DoNotOptimize(result.matrix() = isometry.matrix() * isometry.matrix());
+    benchmark::ClobberMemory();
+  }
+}
 // Benchmark time to multiply an Eigen::Isometry3d with an Eigen::Isometry3d.
 static void multiplyIsometryTimesIsometry(benchmark::State& st)
 {
-  int n_iters = st.range(0);
   Eigen::Isometry3d isometry = createTestIsometry();
+  Eigen::Isometry3d result;
   for (auto _ : st)
   {
-    for (int i = 0; i < n_iters; ++i)
-    {
-      Eigen::Isometry3d result;
-      benchmark::DoNotOptimize(result = isometry * isometry);
-      benchmark::ClobberMemory();
-    }
+    benchmark::DoNotOptimize(result = isometry * isometry);
+    benchmark::ClobberMemory();
   }
 }
 
 // Benchmark time to invert an Eigen::Isometry3d.
 static void inverseIsometry3d(benchmark::State& st)
 {
-  int n_iters = st.range(0);
   Eigen::Isometry3d isometry = createTestIsometry();
+  Eigen::Isometry3d result;
   for (auto _ : st)
   {
-    for (int i = 0; i < n_iters; ++i)
-    {
-      Eigen::Isometry3d result;
-      benchmark::DoNotOptimize(result = isometry.inverse());
-      benchmark::ClobberMemory();
-    }
+    benchmark::DoNotOptimize(result = isometry.inverse());
+    benchmark::ClobberMemory();
   }
 }
 
 // Benchmark time to invert an Eigen::Affine3d(Eigen::Isometry).
 static void inverseAffineIsometry(benchmark::State& st)
 {
-  int n_iters = st.range(0);
-  Eigen::Isometry3d isometry = createTestIsometry();
-  Eigen::Affine3d affine;
-  affine.matrix() = isometry.matrix();
-
+  Eigen::Affine3d affine = createTestIsometry();
+  Eigen::Affine3d result;
   for (auto _ : st)
   {
-    for (int i = 0; i < n_iters; ++i)
-    {
-      Eigen::Affine3d result;
-      benchmark::DoNotOptimize(result = affine.inverse(Eigen::Isometry).affine());
-      benchmark::ClobberMemory();
-    }
+    benchmark::DoNotOptimize(result = affine.inverse(Eigen::Isometry));
+    benchmark::ClobberMemory();
   }
 }
 
 // Benchmark time to invert an Eigen::Affine3d.
 static void inverseAffine(benchmark::State& st)
 {
-  int n_iters = st.range(0);
-  Eigen::Isometry3d isometry = createTestIsometry();
-  Eigen::Affine3d affine;
-  affine.matrix() = isometry.matrix();
-
+  Eigen::Affine3d affine = createTestIsometry();
+  Eigen::Affine3d result;
   for (auto _ : st)
   {
-    for (int i = 0; i < n_iters; ++i)
-    {
-      Eigen::Affine3d result;
-      benchmark::DoNotOptimize(result = affine.inverse().affine());
-      benchmark::ClobberMemory();
-    }
+    benchmark::DoNotOptimize(result = affine.inverse());
+    benchmark::ClobberMemory();
   }
 }
 
 // Benchmark time to invert an Eigen::Matrix4d.
 static void inverseMatrix4d(benchmark::State& st)
 {
-  int n_iters = st.range(0);
-  Eigen::Isometry3d isometry = createTestIsometry();
-  Eigen::Affine3d affine;
-  affine.matrix() = isometry.matrix();
-
+  Eigen::Affine3d affine = createTestIsometry();
+  Eigen::Affine3d result;
   for (auto _ : st)
```

Comment:
Yes, I think that there is some compiler optimization going on. However, I didn't manage to fix that using various variants of `DoNotOptimize`. Finally I had to move on to other stuff...

Reply:
The internal loop doesn't contribute anything when using Google's benchmark lib,
which does exactly that: performing the same operation in a loop for good statistics.
I can imagine that affine inverse is slower, because it cannot optimize memory allocation (needs to skip entries of last, fixed matrix row).